enh: Add evaluate-cardiac-motion tool

baiwenjia opened the pull request:
Hi Andreas,

Thanks a lot for your help. After a second thought, I decided that perhaps I will add a separate file evaluate-cardiac-motion.cc to implement the cardiac displacement and strain analysis. The reasons are below:

1. It would be easier in future if we want to add something to this separate file to analyse displacement and strain in a time series (input can include a sequence of transformation dof files). Such analysis may be specific for cardiac imaging but not for brain imaging.

2. Another reason is that perhaps we could add an option of computing mean strain values for each AHA segment to this separate file, which again may be specific for cardiac imaging.

In case we need a general strain analysis tool in future, I would be happy to add some code to evaluate-distortion.cc to compute strain in a general tangential coordinate system. What do you think?

Many thanks,
Wenjia

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="35" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/biomedia/mirtk/333)
<!-- Reviewable:end -->

schuhschuh commented:
Hi Wenjia,

I think you make valid points and I agree it is best if there is a tool dedicated to cardiac strain and motion measurements. We may add a generic strain computation to `evaluate-distortion` later. Before I would see what Emma is doing regarding strain measurements in the context of brain surface registration.

Your Git commit unfortunately contains some file mode changes of otherwise unmodified source files (`Applications/src/convert-mris.py`, `Applications/src/evaluate-overlap.cc`, and `Applications/src/help-rst.py`). Could you please reset the commit (**no** `--hard` option!) and re-commit only the file `evaluate-cardiac-motion.cc` and changes to `CMakeLists.txt`? I will fix the file mode of `evaluate-overlap.cc` if necessary in another commit which is not related to the new feature here.

Thanks,
Andreas

schuhschuh commented:
Review status: 0 of 2 files reviewed at latest revision, 13 unresolved discussions.

---

*[Applications/src/evaluate-cardiac-motion.cc, line 6 \[r1\]](https://reviewable.io:443/reviews/biomedia/mirtk/333#-KLGrJYVr_x8cJfB5Vw0:-KLGrJYVr_x8cJfB5Vw1:-254411368) ([raw file](https://github.com/biomedia/mirtk/blob/9bc1377791c850beb88e399e9e75168960f44052/Applications/src/evaluate-cardiac-motion.cc#L6)):*
> ```cpp
>  * Copyright 2008-2015 Imperial College London
>  * Copyright 2008-2013 Daniel Rueckert, Julia Schnabel
>  * Copyright 2013-2015 Andreas Schuh
> ```

Please replace the above copyright statement by

Copyright 2016 Imperial College London
Copyright 2016 Wenjia Bai

---

*[Applications/src/evaluate-cardiac-motion.cc, line 29 \[r1\]](https://reviewable.io:443/reviews/biomedia/mirtk/333#-KLGrbOiaQMWoqJbmx_U:-KLGrbOiaQMWoqJbmx_V:275000974) ([raw file](https://github.com/biomedia/mirtk/blob/9bc1377791c850beb88e399e9e75168960f44052/Applications/src/evaluate-cardiac-motion.cc#L29)):*
> ```cpp
> 
> #include "vtkSmartPointer.h"
> #include "vtkPolyDataReader.h"
> ```

Please use the `ReadPolyData` and `WritePolyData` functions declared in `mirtk/PointSetIO.h` instead of `vtkPolyDataReader` and `vtkPolyDataWriter`. These functions are able to write more than just the old legacy VTK file format.

---

*[Applications/src/evaluate-cardiac-motion.cc, line 33 \[r1\]](https://reviewable.io:443/reviews/biomedia/mirtk/333#-KLGrp_i6sbHvSFMk9Fe:-KLGrp_i6sbHvSFMk9Ff:-1450026457) ([raw file](https://github.com/biomedia/mirtk/blob/9bc1377791c850beb88e399e9e75168960f44052/Applications/src/evaluate-cardiac-motion.cc#L33)):*
> ```cpp
> #include "vtkPointData.h"
> #include "vtkFloatArray.h"
> #include "vtkMath.h"
> ```

Please remove this include and instead add `#include "mirtk/VtkMath.h"` before all the VTK header file includes. This is necessary to circumvent a bug in the VTK 6 system installation on Ubuntu 14.

---

*[Applications/src/evaluate-cardiac-motion.cc, line 60 \[r1\]](https://reviewable.io:443/reviews/biomedia/mirtk/333#-KLGs6NpYWit-BZ1BG-L:-KLGs6NpYWit-BZ1BG-M:490209716) ([raw file](https://github.com/biomedia/mirtk/blob/9bc1377791c850beb88e399e9e75168960f44052/Applications/src/evaluate-cardiac-motion.cc#L60)):*
> ```cpp
>   cout << "  mesh:     input myocardial mesh." << endl;
>   cout << "  T:        transformation with regard to the reference frame (end-diastolic frame)." << endl;
>   cout << "  mesh_out: output mesh which stores the vertex-wise motion data." << endl;
> ```

Please remove the `:` and make sure to separate argument name from description by at least 3 space characters. This is required by the `help-rst` tool which is used to markup the help output using reStructuredText for the generation of the online help page. Also, start the description with a capital letter.

Overall, I think it would be better to use option arguments instead of positional arguments. This is more verbose and mistakes of using the command are less likely. It also makes it easier to extend the tool later where possibly not all of these arguments are required.

In particular, I would leave as positional arguments: `<input_mesh> <output_mesh>` and add options `-dofin, -dof <file>` and `-target <file>`. The `,` means alternative option names, i.e.,
```c++
if (OPTION("-dofin") || OPTION("-dof")) dof_name = ARGUMENT;
```

---

*[Applications/src/evaluate-cardiac-motion.cc, line 63 \[r1\]](https://reviewable.io:443/reviews/biomedia/mirtk/333#-KLGt-l8n_pXnNO9Dgi8:-KLGt-l8n_pXnNO9Dgi9:1506442600) ([raw file](https://github.com/biomedia/mirtk/blob/9bc1377791c850beb88e399e9e75168960f44052/Applications/src/evaluate-cardiac-motion.cc#L63)):*
> ```cpp
>   cout << endl;
>   cout << "Optional arguments:" << endl;
>   cout << "  -invert-zaxis      Use this option if the longitudinal direction is the inverted z-axis of the image. (Default: off)" << endl;
> ```

Minor comment, please use `(default: ...)`, i.e., all lowercase here.

---

*[Applications/src/evaluate-cardiac-motion.cc, line 82 \[r1\]](https://reviewable.io:443/reviews/biomedia/mirtk/333#-KLGtBch4NFcFCCQqGg2:-KLGtBch4NFcFCCQqGg3:372733328) ([raw file](https://github.com/biomedia/mirtk/blob/9bc1377791c850beb88e399e9e75168960f44052/Applications/src/evaluate-cardiac-motion.cc#L82)):*
> ```cpp
> // [2] An Elen et al. Three-Dimensional Cardiac Strain Estimation Using Spatio-Temporal Elastic Registration of Ultrasound Images: A Feasibility Study. IEEE Transactions on Medical Imaging, 2008.
> void DetermineLocalCoordinateSystem(vtkSmartPointer<vtkPolyData> mesh, ImageAttributes attr, bool invert_zaxis,
> 				    vtkSmartPointer<vtkDataArray> dir_radial,
> ```

Please replace all tab characters by space characters.

---

*[Applications/src/evaluate-cardiac-motion.cc, line 193 \[r1\]](https://reviewable.io:443/reviews/biomedia/mirtk/333#-KLGtJ7E2aYrXY40EH3O:-KLGtJ7E2aYrXY40EH3P:393360035) ([raw file](https://github.com/biomedia/mirtk/blob/9bc1377791c850beb88e399e9e75168960f44052/Applications/src/evaluate-cardiac-motion.cc#L193)):*
> ```cpp
> int main(int argc, char **argv)
> {
>   REQUIRES_POSARGS(4);
> ```

The `REQUIRES_POSARGS` only ensures that the users supplies as least as many arguments, but may optionally provide more than 4 arguments. As this tool does not have an optional number of positional arguments, use `EXPECTS_POSARGS` instead. It will show the help when the user provides not exactly the specified number of arguments (i.e., either more or less).

---

*[Applications/src/evaluate-cardiac-motion.cc, line 216 \[r1\]](https://reviewable.io:443/reviews/biomedia/mirtk/333#-KLGvkaazYhWeAwjega8:-KLGvkaazYhWeAwjega9:-1601542287) ([raw file](https://github.com/biomedia/mirtk/blob/9bc1377791c850beb88e399e9e75168960f44052/Applications/src/evaluate-cardiac-motion.cc#L216)):*
> ```cpp
>       save_local_coord = true;
>     }
>     else {
> ```

Please add output file format option:
```c++
// [...]
bool compress = true;
int  ascii    = -1;

for (ALL_OPTIONS) {
  // [...]
  else if (OPTION("-ascii")   || OPTION("-nobinary")) ascii = 1;
  else if (OPTION("-noascii") || OPTION("-binary")) ascii = 0;
  else if (OPTION("-compress")) compress = true;
  else if (OPTION("-nocompress")) compress = false;
  // [...]
}
```

---

*[Applications/src/evaluate-cardiac-motion.cc, line 223 \[r1\]](https://reviewable.io:443/reviews/biomedia/mirtk/333#-KLGtjidrg2w5s_I37Wf:-KLGtjidrg2w5s_I37Wg:-1598366003) ([raw file](https://github.com/biomedia/mirtk/blob/9bc1377791c850beb88e399e9e75168960f44052/Applications/src/evaluate-cardiac-motion.cc#L223)):*
> ```cpp
>   // Read input image
>   InitializeIOLibrary();
>   GreyImage image(image_name);
> ```

As the image instance is not used later on, I suggest to place this in a block such that the memory is released immediately after the attributes were obtained:
```c++
ImageAttributes attr;
{
  if (verbose) cout << "Reading image attributes from " << image_name << "...";
  InitializeIOLibrary();
  GreyImage image(image_name);
  attr = image.Attributes();
  if (verbose) cout << " done" << endl;
}
```

---

*[Applications/src/evaluate-cardiac-motion.cc, line 229 \[r1\]](https://reviewable.io:443/reviews/biomedia/mirtk/333#-KLGuEMyeA9FlgJJWqe1:-KLGuEMyeA9FlgJJWqe2:-802583931) ([raw file](https://github.com/biomedia/mirtk/blob/9bc1377791c850beb88e399e9e75168960f44052/Applications/src/evaluate-cardiac-motion.cc#L229)):*
> ```cpp
> 
>   // Read input mesh
>   vtkSmartPointer<vtkPolyDataReader> reader = vtkSmartPointer<vtkPolyDataReader>::New();
> ```

Two comments:
1. The verbose `Reading..` output always follows the actual action instead of coming first.
2. Use `ReadPolyData` function. It will by default exit with error message when file could not be read. It also is able to read more formats supported by VTK such as the new XML based format `.vtp` with ZLIB compression support.

```c++
int mesh_ftype;
vtkSmartPointer<vtkPolyData> mesh = ReadPolyData(mesh_name, &mesh_ftype);
if (ascii == -1) ascii = (mesh_ftype == VTK_ASCII ? 1 : 0);
```

---

*[Applications/src/evaluate-cardiac-motion.cc, line 239 \[r1\]](https://reviewable.io:443/reviews/biomedia/mirtk/333#-KLGugoUSXzRlTikeIMG:-KLGugoUSXzRlTikeIMH:-565993363) ([raw file](https://github.com/biomedia/mirtk/blob/9bc1377791c850beb88e399e9e75168960f44052/Applications/src/evaluate-cardiac-motion.cc#L239)):*
> ```cpp
> 
>   // Read input transformation
>   Transformation *T = dynamic_cast<Transformation *>(Transformation::New(T_name));
> ```

Please use `UniquePtr<Transformation>` to avoid memory leaks.

---

*[Applications/src/evaluate-cardiac-motion.cc, line 277 \[r1\]](https://reviewable.io:443/reviews/biomedia/mirtk/333#-KLGuqhIfwEudDKGqgwa:-KLGuqhIfwEudDKGqgwb:717871955) ([raw file](https://github.com/biomedia/mirtk/blob/9bc1377791c850beb88e399e9e75168960f44052/Applications/src/evaluate-cardiac-motion.cc#L277)):*
> ```cpp
> 
>   // For each point on the mesh
>   for (int i = 0; i < n_points; i++) {
> ```

You could very easily implement this with TBB `parallel_for` for multi-threaded execution. If you would like to see an example of how, let me know. We can write a skeleton in just a minute together.

---

*[Applications/src/evaluate-cardiac-motion.cc, line 365 \[r1\]](https://reviewable.io:443/reviews/biomedia/mirtk/333#-KLGv4fkopqfJCt-bMpt:-KLGv4fkopqfJCt-bMpu:-1055401728) ([raw file](https://github.com/biomedia/mirtk/blob/9bc1377791c850beb88e399e9e75168960f44052/Applications/src/evaluate-cardiac-motion.cc#L365)):*
> ```cpp
>   // Write the output mesh
>   if (verbose) { cout << "Writing mesh to " << mesh_out_name << endl; }
>   vtkPolyDataWriter *writer = vtkPolyDataWriter::New();
> ```

Replace this by
```c++
if (!WritePolyData(mesh_out_name, mesh, compress, static_cast<bool>(ascii))) {
  FatalError("Failed to write output mesh to file " << mesh_out_name);
}
```

The boolean `compress` flag is set by the `-compress` / `-nocompress` options.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/biomedia/mirtk/333)*
<!-- Sent from Reviewable.io -->

schuhschuh commented:
Reviewed 2 of 2 files at r1.
Review status: all files reviewed at latest revision, 13 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/biomedia/mirtk/333)*
<!-- Sent from Reviewable.io -->

baiwenjia commented:
Review status: 1 of 2 files reviewed at latest revision, 13 unresolved discussions.

---

*[Applications/src/evaluate-cardiac-motion.cc, line 6 \[r1\]](https://reviewable.io:443/reviews/biomedia/mirtk/333#-KLGrJYVr_x8cJfB5Vw0:-KLLgzAkNGZu9X2r4Lrg:-499146009) ([raw file](https://github.com/biomedia/mirtk/blob/68868b5c7254157e3170ab9901ed1d96c7aa5fef/Applications/src/evaluate-cardiac-motion.cc#L6)):*
<details><summary><i>Previously, schuhschuh (Andreas Schuh) wrote…</i></summary>
> Please replace the above copyright statement by
> 
> Copyright 2016 Imperial College London
> Copyright 2016 Wenjia Bai
</details>
Done.

---

*[Applications/src/evaluate-cardiac-motion.cc, line 29 \[r1\]](https://reviewable.io:443/reviews/biomedia/mirtk/333#-KLGrbOiaQMWoqJbmx_U:-KLLgzs0l6sVtssjvIKl:-499146009) ([raw file](https://github.com/biomedia/mirtk/blob/68868b5c7254157e3170ab9901ed1d96c7aa5fef/Applications/src/evaluate-cardiac-motion.cc#L29)):*
<details><summary><i>Previously, schuhschuh (Andreas Schuh) wrote…</i></summary>
> Please use the `ReadPolyData` and `WritePolyData` functions declared in `mirtk/PointSetIO.h` instead of `vtkPolyDataReader` and `vtkPolyDataWriter`. These functions are able to write more than just the old legacy VTK file format.
</details>
Done.

---

*[Applications/src/evaluate-cardiac-motion.cc, line 33 \[r1\]](https://reviewable.io:443/reviews/biomedia/mirtk/333#-KLGrp_i6sbHvSFMk9Fe:-KLLh-Rh4wgKqTvWR7Ck:-499146009) ([raw file](https://github.com/biomedia/mirtk/blob/68868b5c7254157e3170ab9901ed1d96c7aa5fef/Applications/src/evaluate-cardiac-motion.cc#L33)):*
<details><summary><i>Previously, schuhschuh (Andreas Schuh) wrote…</i></summary>
> Please remove this include and instead add `#include "mirtk/VtkMath.h"` before all the VTK header file includes. This is necessary to circumvent a bug in the VTK 6 system installation on Ubuntu 14.
</details>
Done.

---

*[Applications/src/evaluate-cardiac-motion.cc, line 60 \[r1\]](https://reviewable.io:443/reviews/biomedia/mirtk/333#-KLGs6NpYWit-BZ1BG-L:-KLLkkSnatkw6X749DOQ:71684668) ([raw file](https://github.com/biomedia/mirtk/blob/68868b5c7254157e3170ab9901ed1d96c7aa5fef/Applications/src/evaluate-cardiac-motion.cc#L60)):*
<details><summary><i>Previously, schuhschuh (Andreas Schuh) wrote…</i></summary>
> Please remove the `:` and make sure to separate argument name from description by at least 3 space characters. This is required by the `help-rst` tool which is used to markup the help output using reStructuredText for the generation of the online help page. Also, start the description with a capital letter.
> 
> Overall, I think it would be better to use option arguments instead of positional arguments. This is more verbose and mistakes of using the command are less likely. It also makes it easier to extend the tool later where possibly not all of these arguments are required.
> 
> In particular, I would leave as positional arguments: `<input_mesh> <output_mesh>` and add options `-dofin, -dof <file>` and `-target <file>`. The `,` means alternative option names, i.e.,
> ```c++
> if (OPTION("-dofin") || OPTION("-dof")) dof_name = ARGUMENT;
> ```
</details>
I re-formatted the arguments as you suggested and added the options "-dof, -dofin" and "-image".

Since my current programme always requires the -dof and -image input, I added Line 214 and Line 229 to check the arguments:

    if (image_name == NULL) { FatalError("Please specify the image name.") }

    if (dof_name == NULL) { FatalError("Please specify the transformation name.") }

I am not sure whether this is appropriate to the MIRTK standard.

---

*[Applications/src/evaluate-cardiac-motion.cc, line 63 \[r1\]](https://reviewable.io:443/reviews/biomedia/mirtk/333#-KLGt-l8n_pXnNO9Dgi8:-KLLhJUWOiAcO8uaT4x8:-499146009) ([raw file](https://github.com/biomedia/mirtk/blob/68868b5c7254157e3170ab9901ed1d96c7aa5fef/Applications/src/evaluate-cardiac-motion.cc#L63)):*
<details><summary><i>Previously, schuhschuh (Andreas Schuh) wrote…</i></summary>
> Minor comment, please use `(default: ...)`, i.e., all lowercase here.
</details>
Done.

---

*[Applications/src/evaluate-cardiac-motion.cc, line 82 \[r1\]](https://reviewable.io:443/reviews/biomedia/mirtk/333#-KLGtBch4NFcFCCQqGg2:-KLLmP2AUsC2AGMb-KGB:-499146009) ([raw file](https://github.com/biomedia/mirtk/blob/68868b5c7254157e3170ab9901ed1d96c7aa5fef/Applications/src/evaluate-cardiac-motion.cc#L82)):*
<details><summary><i>Previously, schuhschuh (Andreas Schuh) wrote…</i></summary>
> Please replace all tab characters by space characters.
</details>
Done.

---

*[Applications/src/evaluate-cardiac-motion.cc, line 193 \[r1\]](https://reviewable.io:443/reviews/biomedia/mirtk/333#-KLGtJ7E2aYrXY40EH3O:-KLLmrOzjyt4ZX9dqVpu:-499146009) ([raw file](https://github.com/biomedia/mirtk/blob/68868b5c7254157e3170ab9901ed1d96c7aa5fef/Applications/src/evaluate-cardiac-motion.cc#L193)):*
<details><summary><i>Previously, schuhschuh (Andreas Schuh) wrote…</i></summary>
> The `REQUIRES_POSARGS` only ensures that the users supplies as least as many arguments, but may optionally provide more than 4 arguments. As this tool does not have an optional number of positional arguments, use `EXPECTS_POSARGS` instead. It will show the help when the user provides not exactly the specified number of arguments (i.e., either more or less).
</details>
Done.

---

*[Applications/src/evaluate-cardiac-motion.cc, line 216 \[r1\]](https://reviewable.io:443/reviews/biomedia/mirtk/333#-KLGvkaazYhWeAwjega8:-KLLqWP_wxerOe4VTotq:-499146009) ([raw file](https://github.com/biomedia/mirtk/blob/68868b5c7254157e3170ab9901ed1d96c7aa5fef/Applications/src/evaluate-cardiac-motion.cc#L216)):*
<details><summary><i>Previously, schuhschuh (Andreas Schuh) wrote…</i></summary>
> Please add output file format option:
> ```c++
> // [...]
> bool compress = true;
> int  ascii    = -1;
> 
> for (ALL_OPTIONS) {
>   // [...]
>   else if (OPTION("-ascii")   || OPTION("-nobinary")) ascii = 1;
>   else if (OPTION("-noascii") || OPTION("-binary")) ascii = 0;
>   else if (OPTION("-compress")) compress = true;
>   else if (OPTION("-nocompress")) compress = false;
>   // [...]
> }
> ```
</details>
Done.

---

*[Applications/src/evaluate-cardiac-motion.cc, line 223 \[r1\]](https://reviewable.io:443/reviews/biomedia/mirtk/333#-KLGtjidrg2w5s_I37Wf:-KLLnHeQMDy3VGLXkXjN:-499146009) ([raw file](https://github.com/biomedia/mirtk/blob/68868b5c7254157e3170ab9901ed1d96c7aa5fef/Applications/src/evaluate-cardiac-motion.cc#L223)):*
<details><summary><i>Previously, schuhschuh (Andreas Schuh) wrote…</i></summary>
> As the image instance is not used later on, I suggest to place this in a block such that the memory is released immediately after the attributes were obtained:
> ```c++
> ImageAttributes attr;
> {
>   if (verbose) cout << "Reading image attributes from " << image_name << "...";
>   InitializeIOLibrary();
>   GreyImage image(image_name);
>   attr = image.Attributes();
>   if (verbose) cout << " done" << endl;
> }
> ```
</details>
Done.

---

*[Applications/src/evaluate-cardiac-motion.cc, line 229 \[r1\]](https://reviewable.io:443/reviews/biomedia/mirtk/333#-KLGuEMyeA9FlgJJWqe1:-KLLnsP1S0tKou6Izcw1:-499146009) ([raw file](https://github.com/biomedia/mirtk/blob/68868b5c7254157e3170ab9901ed1d96c7aa5fef/Applications/src/evaluate-cardiac-motion.cc#L229)):*
<details><summary><i>Previously, schuhschuh (Andreas Schuh) wrote…</i></summary>
> Two comments:
> 1. The verbose `Reading..` output always follows the actual action instead of coming first.
> 2. Use `ReadPolyData` function. It will by default exit with error message when file could not be read. It also is able to read more formats supported by VTK such as the new XML based format `.vtp` with ZLIB compression support.
> 
> ```c++
> int mesh_ftype;
> vtkSmartPointer<vtkPolyData> mesh = ReadPolyData(mesh_name, &mesh_ftype);
> if (ascii == -1) ascii = (mesh_ftype == VTK_ASCII ? 1 : 0);
> ```
</details>
Done.

---

*[Applications/src/evaluate-cardiac-motion.cc, line 239 \[r1\]](https://reviewable.io:443/reviews/biomedia/mirtk/333#-KLGugoUSXzRlTikeIMG:-KLLoYAGMRKheHfDqQZq:-499146009) ([raw file](https://github.com/biomedia/mirtk/blob/68868b5c7254157e3170ab9901ed1d96c7aa5fef/Applications/src/evaluate-cardiac-motion.cc#L239)):*
<details><summary><i>Previously, schuhschuh (Andreas Schuh) wrote…</i></summary>
> Please use `UniquePtr<Transformation>` to avoid memory leaks.
</details>
Done.

---

*[Applications/src/evaluate-cardiac-motion.cc, line 365 \[r1\]](https://reviewable.io:443/reviews/biomedia/mirtk/333#-KLGv4fkopqfJCt-bMpt:-KLLoz_4O4HWk7q7BWn8:-499146009) ([raw file](https://github.com/biomedia/mirtk/blob/68868b5c7254157e3170ab9901ed1d96c7aa5fef/Applications/src/evaluate-cardiac-motion.cc#L365)):*
<details><summary><i>Previously, schuhschuh (Andreas Schuh) wrote…</i></summary>
> Replace this by
> ```c++
> if (!WritePolyData(mesh_out_name, mesh, compress, static_cast<bool>(ascii))) {
>   FatalError("Failed to write output mesh to file " << mesh_out_name);
> }
> ```
> 
> The boolean `compress` flag is set by the `-compress` / `-nocompress` options.
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/biomedia/mirtk/333)*
<!-- Sent from Reviewable.io -->

schuhschuh commented:
Reviewed 1 of 1 files at r2.
Review status: all files reviewed at latest revision, 5 unresolved discussions.

---

*[Applications/src/evaluate-cardiac-motion.cc, line 29 \[r1\]](https://reviewable.io:443/reviews/biomedia/mirtk/333#-KLGrbOiaQMWoqJbmx_U:-KLMEDhebf1-YUpuEeru:1743880875) ([raw file](https://github.com/biomedia/mirtk/blob/68868b5c7254157e3170ab9901ed1d96c7aa5fef/Applications/src/evaluate-cardiac-motion.cc#L29)):*
<details><summary><i>Previously, baiwenjia wrote…</i></summary>
> Done.
</details>
Please also remove the no longer needed includes of `vtkPolyDataReader.h` and `vtkPolyDataWriter.h`.

---

*[Applications/src/evaluate-cardiac-motion.cc, line 60 \[r1\]](https://reviewable.io:443/reviews/biomedia/mirtk/333#-KLGs6NpYWit-BZ1BG-L:-KLMEY6VxFxZQtdcAiae:-1404821882) ([raw file](https://github.com/biomedia/mirtk/blob/68868b5c7254157e3170ab9901ed1d96c7aa5fef/Applications/src/evaluate-cardiac-motion.cc#L60)):*
<details><summary><i>Previously, baiwenjia wrote…</i></summary>
> I re-formatted the arguments as you suggested and added the options "-dof, -dofin" and "-image".
> 
> Since my current programme always requires the -dof and -image input, I added Line 214 and Line 229 to check the arguments:
> 
>     if (image_name == NULL) { FatalError("Please specify the image name.") }
> 
>     if (dof_name == NULL) { FatalError("Please specify the transformation name.") }
> 
> I am not sure whether this is appropriate to the MIRTK standard.
> 
</details>
Yes, that's what you have to do. Also arguments with an option name in front may be required. Not only positional arguments. In the error message, I would include the name of the missing option, i.e.,
```c++
if (!image_name) {
  FatalError("Input -image file name required!");
}
```

In case of this program, I would not require `-image`. If it is not provided, simply assume the identity matrix for the world to image orientation matrix, i.e., simply use the default values of the `ImageAttributes attr` structure.

---

*[Applications/src/evaluate-cardiac-motion.cc, line 223 \[r1\]](https://reviewable.io:443/reviews/biomedia/mirtk/333#-KLGtjidrg2w5s_I37Wf:-KLMFGgGfMdcNHT17Pjr:836150679) ([raw file](https://github.com/biomedia/mirtk/blob/68868b5c7254157e3170ab9901ed1d96c7aa5fef/Applications/src/evaluate-cardiac-motion.cc#L223)):*
<details><summary><i>Previously, baiwenjia wrote…</i></summary>
> Done.
</details>
As noted above, an input `-image` need not be required (even if you always want to use one...). Simply change this to:
```c++
ImageAttributes attr;
if (image_name) {
  // [...]
}
```
and just use `attr` as defined by the default constructor of `ImageAttributes` when `image_name == nullptr`.

---

*[Applications/src/evaluate-cardiac-motion.cc, line 277 \[r1\]](https://reviewable.io:443/reviews/biomedia/mirtk/333#-KLGuqhIfwEudDKGqgwa:-KLMG42s5vhsw5duqFTQ:-1230470987) ([raw file](https://github.com/biomedia/mirtk/blob/68868b5c7254157e3170ab9901ed1d96c7aa5fef/Applications/src/evaluate-cardiac-motion.cc#L277)):*
<details><summary><i>Previously, schuhschuh (Andreas Schuh) wrote…</i></summary>
> You could very easily implement this with TBB `parallel_for` for multi-threaded execution. If you would like to see an example of how, let me know. We can write a skeleton in just a minute together.
</details>
I think it would be nice to make this multi-threaded. You can use the following template code, where the entire block of your `for` loop just goes into the `operator()` definition below:

```c++
class CalculateStrain
{
  vtkPoints            *_Points;
  const Transformation *_Transform;
  vtkDataArray         *_RadialDir;
  vtkDataArray         *_LongitDir;
  vtkDataArray         *_CircumDir;
  vtkDataArray         *_Strain;
  vtkDataArray         *_Displacement;
  vtkPoints            *_TransformedPoints;

  void operator ()(const blocked_range<vtkIdType> ptIds) const
  {
    double p[3], q[3], d[3], radial[3], longit[3], circum[3];
    Matrix J(3, 3);

    for (vtkIdType i = ptIds.begin(); i != ptIds.end(); ++i) {
      // Get surface point coordinates
      _Points->GetPoint(i, p);

      // Get direction vectors of local coordinate axes
      _RadialDir->GetTuple(i, radial);
      _LongitDir->GetTuple(i, longit);
      _CircumDir->GetTuple(i, circum);

      // Transform point and calculate local displacement vectors
      if (_Displacement || _TransformedPoints) {
        q[0] = p[0], q[1] = p[1], q[2] = p[2];
        _Transform->Transform(q[0], q[1], q[2]);
        if (_TransformedPoints) {
          _TransformedPoints->SetPoint(i, q);
        }
        if (_Displacement) {
          vtkMath::Subtract(q, p, d);
          _Displacement->SetComponent(i, 0, vtkMath::Dot(d, radial));
          _Displacement->SetComponent(i, 1, vtkMath::Dot(d, longit));
          _Displacement->SetComponent(i, 2, vtkMath::Dot(d, circum));
        }
      }

      // Calculate strain along local coordinate axes
      if (_Strain) {
        _Transform->Jacobian(J, p[0], p[1], p[2]);
        // [...] rest of strain computation
        _Strain->SetComponent(i, 0, E_l(0,0));
        _Strain->SetComponent(i, 1, E_l(1,1));
        _Strain->SetComponent(i, 2, E_l(2,2));
      }
    }
  }
};


int main(int argc, char *argc[])
{
  // [...]

  // Replace your for loop by
  CalculateStrainAndDisplacement eval;
  eval._Points            = mesh->GetPoints();
  eval._Transform         = T.get();
  eval._RadialDir         = dir_radial;
  eval._LongitDir         = dir_longit;
  eval._CircumDir         = dir_circum;
  eval._Strain            = (compute_strain ? strain : nullptr);
  eval._Displacement      = (compute_disp ? disp : nullptr);
  eval._TransformedPoints = mesh->GetPoints();
  parallel_for(blocked_range<vtkIdType>(0, mesh->GetNumberOfPoints()), eval);
}
```

---

*[Applications/src/evaluate-cardiac-motion.cc, line 50 \[r2\]](https://reviewable.io:443/reviews/biomedia/mirtk/333#-KLMEO5sAhKqbNfp2azO:-KLMEO5sAhKqbNfp2azP:1604656884) ([raw file](https://github.com/biomedia/mirtk/blob/b4a9a59150ea455dd5a8aa1d20d4c22f820062a5/Applications/src/evaluate-cardiac-motion.cc#L50)):*
> ```cpp
> {
>   cout << endl;
>   cout << "Usage: " << name << " <input_mesh> <output_mesh> -image <file> -dof, -dofin <file>" << endl;
> ```

In this example usage, I would only use  `-dof`.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/biomedia/mirtk/333)*
<!-- Sent from Reviewable.io -->

schuhschuh commented:
Review status: all files reviewed at latest revision, 5 unresolved discussions.

---

*[Applications/src/evaluate-cardiac-motion.cc, line 277 \[r1\]](https://reviewable.io:443/reviews/biomedia/mirtk/333#-KLGuqhIfwEudDKGqgwa:-KLML5L4HTOmoBay6p_W:928503938) ([raw file](https://github.com/biomedia/mirtk/blob/68868b5c7254157e3170ab9901ed1d96c7aa5fef/Applications/src/evaluate-cardiac-motion.cc#L277)):*
> void operator ()(const blocked_range<vtkIdType> ptIds) const

This should be better `(const blocked_range<vtkIdType> &ptIds)`, i.e., pass by reference, not value...

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/biomedia/mirtk/333)*
<!-- Sent from Reviewable.io -->

schuhschuh commented:
Review status: all files reviewed at latest revision, 5 unresolved discussions.

---

*[Applications/src/evaluate-cardiac-motion.cc, line 277 \[r1\]](https://reviewable.io:443/reviews/biomedia/mirtk/333#-KLGuqhIfwEudDKGqgwa:-KLMLS2KObOh-TVHsjcj:-1659674053) ([raw file](https://github.com/biomedia/mirtk/blob/68868b5c7254157e3170ab9901ed1d96c7aa5fef/Applications/src/evaluate-cardiac-motion.cc#L277)):*
> eval._Strain            = (compute_strain ? strain : nullptr);
> eval._Displacement      = (compute_disp ? disp : nullptr);

Here you should actually just have `strain` and `disp` which are either NULL or a  valid data array if the output is requested, i.e., before the parallel_for:

```c++
vtkSmartPointer<vtkDataArray> strain;
if (compute_strain) {
  strain = vtkSmartPointer<vtkFloatArray>::New();
  strain->SetName("Strain");
  strain->SetNumberOfComponents(3);
  strain->SetNumberOfTuples(mesh->GetNumberOfPoints());
  mesh->GetPointData()->AddArray(strain);
}
```

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/biomedia/mirtk/333)*
<!-- Sent from Reviewable.io -->

baiwenjia commented:
Review status: all files reviewed at latest revision, 5 unresolved discussions.

---

*[Applications/src/evaluate-cardiac-motion.cc, line 29 \[r1\]](https://reviewable.io:443/reviews/biomedia/mirtk/333#-KLGrbOiaQMWoqJbmx_U:-KLMSrTHLl8621iUDOx8:-499146009) ([raw file](https://github.com/biomedia/mirtk/blob/68868b5c7254157e3170ab9901ed1d96c7aa5fef/Applications/src/evaluate-cardiac-motion.cc#L29)):*
<details><summary><i>Previously, schuhschuh (Andreas Schuh) wrote…</i></summary>
> Please also remove the no longer needed includes of `vtkPolyDataReader.h` and `vtkPolyDataWriter.h`.
</details>
Done.

---

*[Applications/src/evaluate-cardiac-motion.cc, line 60 \[r1\]](https://reviewable.io:443/reviews/biomedia/mirtk/333#-KLGs6NpYWit-BZ1BG-L:-KLMTRVD2kb3gwUt-sB7:-499146009) ([raw file](https://github.com/biomedia/mirtk/blob/68868b5c7254157e3170ab9901ed1d96c7aa5fef/Applications/src/evaluate-cardiac-motion.cc#L60)):*
<details><summary><i>Previously, schuhschuh (Andreas Schuh) wrote…</i></summary>
> Yes, that's what you have to do. Also arguments with an option name in front may be required. Not only positional arguments. In the error message, I would include the name of the missing option, i.e.,
> ```c++
> if (!image_name) {
>   FatalError("Input -image file name required!");
> }
> ```
> 
> In case of this program, I would not require `-image`. If it is not provided, simply assume the identity matrix for the world to image orientation matrix, i.e., simply use the default values of the `ImageAttributes attr` structure.
</details>
Done.

---

*[Applications/src/evaluate-cardiac-motion.cc, line 223 \[r1\]](https://reviewable.io:443/reviews/biomedia/mirtk/333#-KLGtjidrg2w5s_I37Wf:-KLMTag6s32CuXvK6TPN:-1228032272) ([raw file](https://github.com/biomedia/mirtk/blob/68868b5c7254157e3170ab9901ed1d96c7aa5fef/Applications/src/evaluate-cardiac-motion.cc#L223)):*
<details><summary><i>Previously, schuhschuh (Andreas Schuh) wrote…</i></summary>
> As noted above, an input `-image` need not be required (even if you always want to use one...). Simply change this to:
> ```c++
> ImageAttributes attr;
> if (image_name) {
>   // [...]
> }
> ```
> and just use `attr` as defined by the default constructor of `ImageAttributes` when `image_name == nullptr`.
</details>
I did not change this since normally cardiac images do not have z-axis = 0, 0, 1 so I want the user to always provide the cardiac image for determining the z-axis and longitudinal direction.

---

*[Applications/src/evaluate-cardiac-motion.cc, line 277 \[r1\]](https://reviewable.io:443/reviews/biomedia/mirtk/333#-KLGuqhIfwEudDKGqgwa:-KLMYosAu6zQmFpaihhN:-1952417819) ([raw file](https://github.com/biomedia/mirtk/blob/68868b5c7254157e3170ab9901ed1d96c7aa5fef/Applications/src/evaluate-cardiac-motion.cc#L277)):*
<details><summary><i>Previously, schuhschuh (Andreas Schuh) wrote…</i></summary>
> > eval._Strain            = (compute_strain ? strain : nullptr);
> > eval._Displacement      = (compute_disp ? disp : nullptr);
> 
> Here you should actually just have `strain` and `disp` which are either NULL or a  valid data array if the output is requested, i.e., before the parallel_for:
> 
> ```c++
> vtkSmartPointer<vtkDataArray> strain;
> if (compute_strain) {
>   strain = vtkSmartPointer<vtkFloatArray>::New();
>   strain->SetName("Strain");
>   strain->SetNumberOfComponents(3);
>   strain->SetNumberOfTuples(mesh->GetNumberOfPoints());
>   mesh->GetPointData()->AddArray(strain);
> }
> ```
</details>
Done.

However, I changed the "class CalculateDisplacementAndStrain" into a struct. Otherwise, the compiler is complaining about access to private members.

---

*[Applications/src/evaluate-cardiac-motion.cc, line 50 \[r2\]](https://reviewable.io:443/reviews/biomedia/mirtk/333#-KLMEO5sAhKqbNfp2azO:-KLMT5iAvES9C27xaNEJ:-499146009) ([raw file](https://github.com/biomedia/mirtk/blob/b4a9a59150ea455dd5a8aa1d20d4c22f820062a5/Applications/src/evaluate-cardiac-motion.cc#L50)):*
<details><summary><i>Previously, schuhschuh (Andreas Schuh) wrote…</i></summary>
> In this example usage, I would only use  `-dof`.
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/biomedia/mirtk/333)*
<!-- Sent from Reviewable.io -->

schuhschuh commented:
Reviewed 1 of 1 files at r3.
Review status: all files reviewed at latest revision, 9 unresolved discussions.

---

*[Applications/src/evaluate-cardiac-motion.cc, line 29 \[r1\]](https://reviewable.io:443/reviews/biomedia/mirtk/333#-KLGrbOiaQMWoqJbmx_U:-KLMbSHQ0U1Yi2Re_8Yn:505600080) ([raw file](https://github.com/biomedia/mirtk/blob/68868b5c7254157e3170ab9901ed1d96c7aa5fef/Applications/src/evaluate-cardiac-motion.cc#L29)):*
<details><summary><i>Previously, baiwenjia wrote…</i></summary>
> Done.
</details>
Thanks, but you still need to remove `#include "vtkPolyDataReader.h"`.

---

*[Applications/src/evaluate-cardiac-motion.cc, line 277 \[r1\]](https://reviewable.io:443/reviews/biomedia/mirtk/333#-KLGuqhIfwEudDKGqgwa:-KLMcupqMLfO87p6uQ4l:314021998) ([raw file](https://github.com/biomedia/mirtk/blob/68868b5c7254157e3170ab9901ed1d96c7aa5fef/Applications/src/evaluate-cardiac-motion.cc#L277)):*
<details><summary><i>Previously, baiwenjia wrote…</i></summary>
> Done.
> 
> However, I changed the "class CalculateDisplacementAndStrain" into a struct. Otherwise, the compiler is complaining about access to private members.
</details>
Yes, `struct` is fine. It is the same as when you use `class` and the `public:` keyword before the member declarations.

---

*[Applications/src/evaluate-cardiac-motion.cc, line 211 \[r2\]](https://reviewable.io:443/reviews/biomedia/mirtk/333#-KLMcAXlGEv4bjeihOjF:-KLMcAXlGEv4bjeihOjG:-293023212) ([raw file](https://github.com/biomedia/mirtk/blob/b4a9a59150ea455dd5a8aa1d20d4c22f820062a5/Applications/src/evaluate-cardiac-motion.cc#L211)):*
> ```cpp
>   for (ALL_OPTIONS) {
>     if      (OPTION("-image"))            { image_name = ARGUMENT; }
>     else if (OPTION("-dof") || OPTION("-dofin")) { dof_name = ARGUMENT; }
> ```

Please revert this change as noted above.

---

*[Applications/src/evaluate-cardiac-motion.cc, line 59 \[r3\]](https://reviewable.io:443/reviews/biomedia/mirtk/333#-KLMb__O9dtxFhGTyo_u:-KLMb__O9dtxFhGTyo_v:1708130540) ([raw file](https://github.com/biomedia/mirtk/blob/6192b6ccf95538d49628015a34cf5dc0e1705dfa/Applications/src/evaluate-cardiac-motion.cc#L59)):*
> ```cpp
>   cout << "  output_mesh          The output mesh which stores the vertex-wise motion data." << endl;
>   cout << "  -image file          Cardiac image whose z-axis defines the longitudinal direction." << endl;
>   cout << "  -dof file            Transformation with regard to the reference frame (end-diastolic frame)." << endl;
> ```

Sorry I was not clear before, I only meant to remove the alternative `-dofin` name from the `usage:` line of the help output. As other MIRTK tools use `-dofin`, it would be good to still keep this alternative option name.

So:
```c++
cout << "Usage: " << name << " <input_mesh> <output_mesh> -image <file> -dof <file>" << endl;
```
But:
```c++
cout << "Arguments:" << endl;
// ...
cout << "Options:" << endl;
// ...
cout << "  -dof, -dofin file            Transformation with regard to the reference frame (end-diastolic frame)." << endl;
```

---

*[Applications/src/evaluate-cardiac-motion.cc, line 317 \[r3\]](https://reviewable.io:443/reviews/biomedia/mirtk/333#-KLMcEnvo1CTbgkC07hn:-KLMcEnvo1CTbgkC07ho:1640226045) ([raw file](https://github.com/biomedia/mirtk/blob/6192b6ccf95538d49628015a34cf5dc0e1705dfa/Applications/src/evaluate-cardiac-motion.cc#L317)):*
> ```cpp
> 
>   // Read input image
>   if (image_name == NULL) { FatalError("Input -image file name required!"); }
> ```

Please use `nullptr` instead.

---

*[Applications/src/evaluate-cardiac-motion.cc, line 338 \[r3\]](https://reviewable.io:443/reviews/biomedia/mirtk/333#-KLMcWxNMJTwZFSHfjTw:-KLMcWxNMJTwZFSHfjTx:-80186773) ([raw file](https://github.com/biomedia/mirtk/blob/6192b6ccf95538d49628015a34cf5dc0e1705dfa/Applications/src/evaluate-cardiac-motion.cc#L338)):*
> ```cpp
> 
>   // Read input transformation
>   if (dof_name == NULL) { FatalError("Input -dof file name required!"); }
> ```

Same here, please use `nullptr`.

---

*[Applications/src/evaluate-cardiac-motion.cc, line 367 \[r3\]](https://reviewable.io:443/reviews/biomedia/mirtk/333#-KLMcZzq5Rco_DUTlNu-:-KLMcZzq5Rco_DUTlNu0:-2049159786) ([raw file](https://github.com/biomedia/mirtk/blob/6192b6ccf95538d49628015a34cf5dc0e1705dfa/Applications/src/evaluate-cardiac-motion.cc#L367)):*
> ```cpp
> 
>   // Prepare the array for storing the displacements and strains
>   vtkSmartPointer<vtkFloatArray> disp = NULL;
> ```

You do not need to explicitly initialize a `vtkSmartPointer`. The default constructor will set the pointer to `0`. Hence, just use:
```c++
vtkSmartPointer<vtkDataArray> disp;
```
Also, don't specify the exact array data type here, only when you instantiate the array.

---

*[Applications/src/evaluate-cardiac-motion.cc, line 375 \[r3\]](https://reviewable.io:443/reviews/biomedia/mirtk/333#-KLMcowmwsr75WPtaFw9:-KLMcowmwsr75WPtaFwA:-1121266108) ([raw file](https://github.com/biomedia/mirtk/blob/6192b6ccf95538d49628015a34cf5dc0e1705dfa/Applications/src/evaluate-cardiac-motion.cc#L375)):*
> ```cpp
>   }
> 
>   vtkSmartPointer<vtkFloatArray> strain = NULL;
> ```

Same that applies to `disp`, please remove the `= NULL`.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/biomedia/mirtk/333)*
<!-- Sent from Reviewable.io -->

baiwenjia commented:
Reviewed 1 of 2 files at r1, 1 of 1 files at r3.
Review status: all files reviewed at latest revision, 8 unresolved discussions.

---

*[Applications/src/evaluate-cardiac-motion.cc, line 29 \[r1\]](https://reviewable.io:443/reviews/biomedia/mirtk/333#-KLGrbOiaQMWoqJbmx_U:-KLQY9rr34dCCal7GwqG:822339607) ([raw file](https://github.com/biomedia/mirtk/blob/68868b5c7254157e3170ab9901ed1d96c7aa5fef/Applications/src/evaluate-cardiac-motion.cc#L29)):*
<details><summary><i>Previously, schuhschuh (Andreas Schuh) wrote…</i></summary>
> Thanks, but you still need to remove `#include "vtkPolyDataReader.h"`.
</details>
Sorry, Andreas. I forgot to mention that if I removed vtkPolyDataReader.h, the compiler would complain that "error: ‘VTK_ASCII’ was not declared in this scope." That is why I kept this line.

My current solution is to replace this line with "#include "vtkDataReader.h"", in which the macro VTK_ASCII is defined.

---

*[Applications/src/evaluate-cardiac-motion.cc, line 277 \[r1\]](https://reviewable.io:443/reviews/biomedia/mirtk/333#-KLGuqhIfwEudDKGqgwa:-KLQa5SNfnNLUAMOWywx:-499146009) ([raw file](https://github.com/biomedia/mirtk/blob/68868b5c7254157e3170ab9901ed1d96c7aa5fef/Applications/src/evaluate-cardiac-motion.cc#L277)):*
<details><summary><i>Previously, schuhschuh (Andreas Schuh) wrote…</i></summary>
> Yes, `struct` is fine. It is the same as when you use `class` and the `public:` keyword before the member declarations.
</details>
Done.

---

*[Applications/src/evaluate-cardiac-motion.cc, line 211 \[r2\]](https://reviewable.io:443/reviews/biomedia/mirtk/333#-KLMcAXlGEv4bjeihOjF:-KLQZj0FzGRFxgzwaDBc:-499146009) ([raw file](https://github.com/biomedia/mirtk/blob/b4a9a59150ea455dd5a8aa1d20d4c22f820062a5/Applications/src/evaluate-cardiac-motion.cc#L211)):*
<details><summary><i>Previously, schuhschuh (Andreas Schuh) wrote…</i></summary>
> Please revert this change as noted above.
</details>
Done.

---

*[Applications/src/evaluate-cardiac-motion.cc, line 59 \[r3\]](https://reviewable.io:443/reviews/biomedia/mirtk/333#-KLMb__O9dtxFhGTyo_u:-KLQYzE1W584BOzM9S8u:-499146009) ([raw file](https://github.com/biomedia/mirtk/blob/6192b6ccf95538d49628015a34cf5dc0e1705dfa/Applications/src/evaluate-cardiac-motion.cc#L59)):*
<details><summary><i>Previously, schuhschuh (Andreas Schuh) wrote…</i></summary>
> Sorry I was not clear before, I only meant to remove the alternative `-dofin` name from the `usage:` line of the help output. As other MIRTK tools use `-dofin`, it would be good to still keep this alternative option name.
> 
> So:
> ```c++
> cout << "Usage: " << name << " <input_mesh> <output_mesh> -image <file> -dof <file>" << endl;
> ```
> But:
> ```c++
> cout << "Arguments:" << endl;
> // ...
> cout << "Options:" << endl;
> // ...
> cout << "  -dof, -dofin file            Transformation with regard to the reference frame (end-diastolic frame)." << endl;
> ```
</details>
Done.

---

*[Applications/src/evaluate-cardiac-motion.cc, line 317 \[r3\]](https://reviewable.io:443/reviews/biomedia/mirtk/333#-KLMcEnvo1CTbgkC07hn:-KLQ__TTqxiDxoKnwxdm:-499146009) ([raw file](https://github.com/biomedia/mirtk/blob/6192b6ccf95538d49628015a34cf5dc0e1705dfa/Applications/src/evaluate-cardiac-motion.cc#L317)):*
<details><summary><i>Previously, schuhschuh (Andreas Schuh) wrote…</i></summary>
> Please use `nullptr` instead.
</details>
Done.

---

*[Applications/src/evaluate-cardiac-motion.cc, line 338 \[r3\]](https://reviewable.io:443/reviews/biomedia/mirtk/333#-KLMcWxNMJTwZFSHfjTw:-KLQ_bYT8ke11ioIaAqD:-499146009) ([raw file](https://github.com/biomedia/mirtk/blob/6192b6ccf95538d49628015a34cf5dc0e1705dfa/Applications/src/evaluate-cardiac-motion.cc#L338)):*
<details><summary><i>Previously, schuhschuh (Andreas Schuh) wrote…</i></summary>
> Same here, please use `nullptr`.
</details>
Done.

---

*[Applications/src/evaluate-cardiac-motion.cc, line 367 \[r3\]](https://reviewable.io:443/reviews/biomedia/mirtk/333#-KLMcZzq5Rco_DUTlNu-:-KLQ_pueDr_tg436zcXS:-499146009) ([raw file](https://github.com/biomedia/mirtk/blob/6192b6ccf95538d49628015a34cf5dc0e1705dfa/Applications/src/evaluate-cardiac-motion.cc#L367)):*
<details><summary><i>Previously, schuhschuh (Andreas Schuh) wrote…</i></summary>
> You do not need to explicitly initialize a `vtkSmartPointer`. The default constructor will set the pointer to `0`. Hence, just use:
> ```c++
> vtkSmartPointer<vtkDataArray> disp;
> ```
> Also, don't specify the exact array data type here, only when you instantiate the array.
</details>
Done.

---

*[Applications/src/evaluate-cardiac-motion.cc, line 375 \[r3\]](https://reviewable.io:443/reviews/biomedia/mirtk/333#-KLMcowmwsr75WPtaFw9:-KLQ_rV_NkwsYaqKg-by:-499146009) ([raw file](https://github.com/biomedia/mirtk/blob/6192b6ccf95538d49628015a34cf5dc0e1705dfa/Applications/src/evaluate-cardiac-motion.cc#L375)):*
<details><summary><i>Previously, schuhschuh (Andreas Schuh) wrote…</i></summary>
> Same that applies to `disp`, please remove the `= NULL`.
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/biomedia/mirtk/333)*
<!-- Sent from Reviewable.io -->

schuhschuh commented:
Reviewed 1 of 1 files at r4.
Review status: all files reviewed at latest revision, 8 unresolved discussions.

---

*[Applications/src/evaluate-cardiac-motion.cc, line 29 \[r1\]](https://reviewable.io:443/reviews/biomedia/mirtk/333#-KLGrbOiaQMWoqJbmx_U:-KLQnWTPEE-tIjWEQUn_:-1584468595) ([raw file](https://github.com/biomedia/mirtk/blob/68868b5c7254157e3170ab9901ed1d96c7aa5fef/Applications/src/evaluate-cardiac-motion.cc#L29)):*
<details><summary><i>Previously, baiwenjia wrote…</i></summary>
> Sorry, Andreas. I forgot to mention that if I removed vtkPolyDataReader.h, the compiler would complain that "error: ‘VTK_ASCII’ was not declared in this scope." That is why I kept this line.
> 
> My current solution is to replace this line with "#include "vtkDataReader.h"", in which the macro VTK_ASCII is defined.
</details>
I see, that makes sense. I will open a separate issue to modify `ReadPolyData` to return a boolean `ascii` flag instead of this `VTK_ASCII` or `VTK_BINARY` instance in this case. I would like programs to only depend on the MIRTK IO library, but not the VTK IO libraries if possible.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/biomedia/mirtk/333)*
<!-- Sent from Reviewable.io -->

schuhschuh commented:
Thanks, @baiwenjia. Looks good now.

schuhschuh commented:
Minor refactoring of the tool as part of #339, see [this commit](https://github.com/BioMedIA/MIRTK/pull/339/commits/cbb74b0776916198629f7673dc07286d317b378f).

baiwenjia commented:
Hi Andreas,

Thanks a lot for refactoring the code and adding the parallel
implementation. I reviewed and tested the code but do not know how to
respond to the message so I created a new pull request.

Best regards,
Wenjia

On 29 June 2016 at 20:14, Andreas Schuh <notifications@github.com> wrote:

> Minor refactoring of the tool as part of #339
> <https://github.com/BioMedIA/MIRTK/pull/339>, see this commit
> <https://github.com/BioMedIA/MIRTK/pull/339/commits/cbb74b0776916198629f7673dc07286d317b378f>
> .
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/BioMedIA/MIRTK/pull/333#issuecomment-229458873>, or mute
> the thread
> <https://github.com/notifications/unsubscribe/AKYSLdN28YaknNoZO2EhqHGczE_bE8l5ks5qQsQcgaJpZM4I_Ccp>
> .
>
